### PR TITLE
VPC-CNI minimal image builds

### PR DESCRIPTION
**Pull request by jdn5126**
**What type of PR is this?**
Enhancement

**Which issue does this PR fix**:
N/A

**What does this PR do / Why do we need it**:
This pull request migrates the VPC-CNI and VPC-CNI init images to use EKS minimal build image. This PR applied the diff from #1726 and updated it to handle recent changes. It was also tested to be compatible with #2137 , but those changes were left commented out as #2137 is not yet complete.

The VPC CNI image is based on `public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base-iptables:latest.2`, which installs `iptables-legacy` and `iptables-nft`, but sets `iptables-legacy` as the default using `update-alternatives`.

The VPC CNI init image is based on `public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base-glibc:latest.2`. The reason I could not use the base image is that the binaries are built in `pie` mode for security reasons, and building with `pie` requires the loader to be loaded despite the binary being static.

I considered splitting this into two PRs: one to add the Golang entrypoints and one to do the move, but concluded that it is easiest to do as one PR, and if there are any issues, it is simple to just revert the move.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:
N/A

**Testing done on this change**:
I manually verified that all integration tests pass against this change. Output for init container:
```
$ kubectl logs aws-node-rg6rm -n kube-system -c aws-vpc-cni-init
time="2022-11-21T19:35:43Z" level=info msg="Copying CNI plugin binaries ..."
time="2022-11-21T19:35:43Z" level=info msg="Copied all CNI plugin binaries to /host/opt/cni/bin\n"
Installed /host/opt/cni/bin/loopback
Installed /host/opt/cni/bin/portmap
Installed /host/opt/cni/bin/bandwidth
Installed /host/opt/cni/bin/aws-cni-support.sh
time="2022-11-21T19:35:43Z" level=info msg="Found primaryMAC 0a:55:51:01:a4:d1"
time="2022-11-21T19:35:43Z" level=info msg="Found primaryIF eth0"
time="2022-11-21T19:35:43Z" level=info msg="Updated net/ipv4/conf/eth0/rp_filter to 2"
time="2022-11-21T19:35:43Z" level=info msg="Updated net/ipv4/tcp_early_demux to 1"
time="2022-11-21T19:35:43Z" level=info msg="CNI init container done"
```

Output for VPC-CNI:
```
$ kubectl logs aws-node-rg6rm -n kube-system -c aws-node
time="2022-11-21T19:35:43Z" level=info msg="Install CNI binaries.."
Installed /host/opt/cni/bin/aws-cni
time="2022-11-21T19:35:43Z" level=info msg="Starting IPAM daemon... "
time="2022-11-21T19:35:43Z" level=info msg="Checking for IPAM connectivity... "
Installed /host/opt/cni/bin/egress-v4-cni
time="2022-11-21T19:35:44Z" level=info msg="Copying config file... "
time="2022-11-21T19:35:44Z" level=info msg="Successfully copied CNI plugin binary and config file."
```

**Automation added to e2e**:
N/A

**Will this PR introduce any new dependencies?**:
No, it will actually reduce dependencies.

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
This will not break upgrades or downgrades. Upgrade and downgrade has been tested with a running cluster.

**Does this change require updates to the CNI daemonset config files to work?**:
No

**Does this PR introduce any user-facing change?**:
The image base is different.

```release-note
Move VPC CNI and VPC CNI init images to use EKS minimal base image.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bwagner5** [comment]
Nice! Do you know the full image size differences?

**Comment by jdn5126** [comment]
> Nice! Do you know the full image size differences?

Comparing upstream (v1.12.0) to my personal ECR:
VPC-CNI Init: 277.24 MB -> 19.90MB
VPC-CNI: 293.37 MB -> 47.67MB

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:55]: we don't need the \n in logs.

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:38]: nit: i prefer to avoid use non-standard abbreviation and simply use `default` :D

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:123]: the original shell-based script have '-e'(Use this option to ignore errors about unknown keys), while Golang version don't. wondering why we have it in the shell version

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:164]: just a thought, why we make it two container and sync using filePath?
how about just one container, that runs the init first then rest? (if we want parallel processing, we can use goroutings)

Comment (M00nF1sh) [reply cmd/aws-vpc-cni-init/main.go:164]: also nit:
since we are using golang, i prefer to split this "main" into smaller functions, such as `findNetLinkByMac`  `configureRPFilter()`, `configureIPv6Settings`

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:38]: `default` is a reserved keyword in golang, so this was abbreviated. Do you think `defaultVal` is better?

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:123]: Ah nice catch, looks like this comes from https://github.com/aws/amazon-vpc-cni-k8s/issues/1241 . Not sure if we still need to support kernels this old anymore, but it seems worth catching this error

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:164]: That's a good point. Now that #2137 is removing the init container, we can actually just merge the two. I thought the init container was separate as it needed escalated privileges for setup, but it looks like that is not the case. I think we should do this in stages, though, i.e. this should merge with init container support, then #2137 can be rebased to merge with a single container.

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [reply cmd/aws-vpc-cni-init/main.go:38]: defaultVal sounds good to me :D      `def ` is more like short for definition

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:164]: Did some more digging, and it looks like we cannot merge the two containers as the `init` container requires privileged  permissions to set `/proc/sys/net/ipv4/conf/eth0/rp_filter` to loose mode.

I see you had a suggestion in https://github.com/aws/amazon-vpc-cni-k8s/issues/796 to use `src_valid_mark` instead, so I am trying that now

Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:164]: Oh nvm, setting that requires the same level of privilege, so yeah privilege is the limitation for the two containers

File: cmd/aws-vpc-cni-init/main.go
Comment (kishorj) [inline cmd/aws-vpc-cni-init/main.go:136]: use sysctl format instead of the path, for example:
net.ipv6.conf.all.disable_ipv6 and do the necessary translation from the sysctl package. It will be easier to correlate with the sysctl documentation.

File: cmd/aws-vpc-cni-init/main.go
Comment (kishorj) [inline cmd/aws-vpc-cni-init/main.go:100]: use a different name - sys sounds confusing - maybe sysctlUtil

File: utils/sysctl/sysctl.go
Comment (kishorj) [inline utils/sysctl/sysctl.go:19]: why not Get and Set instead?

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:182]: In the latest diff, I broke up `aws-vpc-cni-init/main.go:main` into several parts. For `aws-vpc-cni/main.go:main`, I think it was already split up pretty well

File: scripts/dockerfiles/Dockerfile.init
Comment (jdn5126) [reply scripts/dockerfiles/Dockerfile.init:21]: I am fine either way, but the reason I was thinking `latest` is ok here is that updates to this base repo are highly scrutinized and rarely happen, so we get the benefit of automatically getting fixes and updates without much stability risk or churn

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: Our `sysctl` library actually writes to that file path, i.e. calls `WriteFile` on `proc/sys/net/ipv6/conf/all/disable_ipv6`, so I think the readability improvement of passing it in dotted format just to string replace back to path format may not be worth the computational overhead. I am fine either way, though

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: will do

File: utils/sysctl/sysctl.go
Comment (jdn5126) [reply utils/sysctl/sysctl.go:19]: No reason, I was just copying what was originally done. Changing to `Get` and `Set` now

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [reply cmd/aws-vpc-cni-init/main.go:182]: Ref for init-container use case - https://github.com/aws/amazon-vpc-cni-k8s/issues/796

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [inline cmd/aws-vpc-cni-init/main.go:186]: Lets delete these lines.

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [inline cmd/aws-vpc-cni-init/main.go:192]: Same as above, lets clean it up.

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:186]: These are going to be uncommented for #2137 , so I was leaving them to make it easier for that PR to rebase

File: cmd/aws-vpc-cni/main.go
Comment (jayanthvn) [inline cmd/aws-vpc-cni/main.go:179]: This function is anyways commented, we can delete it.

Comment (jdn5126) [reply cmd/aws-vpc-cni/main.go:179]: same as above, this is temporarily commented to simplify rebase for #2137

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [reply cmd/aws-vpc-cni-init/main.go:186]: I see, sure.

File: scripts/dockerfiles/Dockerfile.init
Comment (jayanthvn) [reply scripts/dockerfiles/Dockerfile.init:21]: Lets keep it as latest similar to what we have today. Anyways for CVEs we do a rebuild and get all the latest packages.

File: cmd/aws-vpc-cni-init/main.go
Comment (bryantbiggs) [reply cmd/aws-vpc-cni-init/main.go:186]: This will make it harder since it would in theory cause a conflict as opposed to a net new addition in the follow on PR

Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:186]: These files do not exist in the new PR, so will not cause any conflict, and the entrypoint scripts the new PR is modifying are deleted by this PR, so will cause a conflict, but that was unavoidable and the resolution is easy (accept delete of file). #2137 needs cleanup anyway before it can be merged, so the goal is for the diff in these files to just be the uncommented lines

Comment (bryantbiggs) [reply cmd/aws-vpc-cni-init/main.go:186]: Ok, that's just a bit odd but ok

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:50]: nit: prefer to just return `""` when an error is returned.

File: cmd/aws-vpc-cni-init/main.go
Comment (M00nF1sh) [inline cmd/aws-vpc-cni-init/main.go:85]: shouldn't this be `err == nil`?

File: cmd/aws-vpc-cni-init/main.go
Comment (achevuru) [inline cmd/aws-vpc-cni-init/main.go]: We need `host-local` plugin for IPv6  - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/misc/10-aws.conflist#L23

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:85]: If the path to `tcp_early_demux` does not exist, we return error. If an error is returned, we do not want to write to `tcp_early_demux`

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:36]: `HOST_CNI_BIN_PATH` was copied from `scripts/init.sh`. The default value is `/host/opt/cni/bin`, but we do currently allow for it to be overridden. We do not advertise this env var, though

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: I do see `host-local` in entrypoint script here: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/scripts/entrypoint.sh#L154 , but it is not present in init script: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/scripts/init.sh#L25 . I assume it should be in both?

Also, it does seem like IPv6 test suite passes without this. I will add it, though.

File: cmd/aws-vpc-cni/main.go
Comment (achevuru) [inline cmd/aws-vpc-cni/main.go:302]: Should this check be paired with `ENABLE_POD_ENI`? Why make this a mandatory check for all modes?

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [reply cmd/aws-vpc-cni-init/main.go]: I think we can keep it only in init script. It is confusing to have it in two places. We can take this chance to cleanup the code.

Comment (jayanthvn) [reply cmd/aws-vpc-cni-init/main.go]: btw @achevuru , some of my code was implemented before v6 hence it is a good callout to test v6 before merging.

Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: but when someone sets `AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER` to true, which is the default, then it is main container that should install this binary

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni/main.go:302]: I was just copying the check from the existing `entrypoint` script, but yeah, makes sense to check this only when `ENABLE_POD_ENI` is set.

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go:85]: Oh, nvm, you are right here. I was not being thorough in testing after this change

File: cmd/aws-vpc-cni/main.go
Comment (achevuru) [inline cmd/aws-vpc-cni/main.go:393]: I wonder if we really need this anymore. v1.6.0+ anyways migrated to `10-aws.conflist` and it has been 2+ years already.

File: cmd/aws-vpc-cni-init/main.go
Comment (jayanthvn) [reply cmd/aws-vpc-cni-init/main.go]: Init container will always be called right? so one place should be fine. I think it was added initially when init container was first supported..

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni/main.go:393]: The only thing we are doing here is deleting this file if it exists. But since no one consumes it and we do not allow upgrades from version < `v1.6.0` to `v1.12.1`, it does seem like this can be deleted. I'll test it out now

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: I guess the question then is whether we even support `AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER` being configured as `"false"`. Currently, both the init container and the main container installs these binaries. Since the init container does it unconditionally, there is no need for the main container to do it either way

Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: nvm, it looks like `AWS_VPC_K8S_CNI_CONFIGURE_RPFILTER` is used here: https://github.com/aws/amazon-vpc-cni-k8s/blob/f21b587fdf6370a01b181ddce797398eb7951a23/pkg/networkutils/network.go#L298 . So really, this env variable does not make sense as a gate for the main entrypoint installing the same binaries that init entrypoint already did

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: fixed in latest revision

File: cmd/aws-vpc-cni-init/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni-init/main.go]: Ran all IPv6 and IPv4 test suites and verified that they passed. Modified entrypoint scripts such that only init container installs external plugins. This also allowed me to simplify dockerfiles.

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [reply cmd/aws-vpc-cni/main.go]: removed in latest revision

**Review by achevuru — APPROVED** [review]

**Comment by bryantbiggs** [comment]
awesome work @jdn5126 🎉

**Pull request merged**
